src/notebooks/controllers/controllerRegistration.ts

```diff
@@ -160,6 +175,68 @@ export class ControllerRegistration implements IControllerRegistration {
         const id = this.getControllerId(metadata, notebookType);
         return this.registeredControllers.get(id);
     }
+    @swallowExceptions()
+    private async swapKernelSpecConnectionControllerWithLiveController(kernel: IKernel) {
+        if (kernel.kernelConnectionMetadata.kind === 'startUsingRemoteKernelSpec') {
+            if (!kernel.session?.kernel?.id) {
+                return;
+            }
+            // Check if we have a controller registered for this live kernel session.
+            // If not, then create one.
+            // Next, then swap the metadata & controller in the kernel.
+            const liveKernelConnection: LiveRemoteKernelConnectionMetadata = {
+                kind: 'connectToLiveRemoteKernel',
+                baseUrl: kernel.kernelConnectionMetadata.baseUrl,
+                id: kernel.session.kernel.id,
+                kernelModel: {
+                    ...kernel.kernelConnectionMetadata.kernelSpec,
+                    model: {
+                        id: kernel.session.kernel.id,
+                        name: kernel.session.kernel.name,
+                        path: path.basename(kernel.uri.path),
+                        type: 'notebook',
+                        kernel: kernel.session.kernel.model
+                    },
+                    notebook: {
+                        path: path.basename(kernel.uri.path)
+                    },
+                    name: kernel.session.kernel.name,
+                    numberOfConnections: 1,
+                    lastActivityTime: new Date(Date.now()),
+                    id: kernel.session.kernel.id
+                },
+                serverId: kernel.kernelConnectionMetadata.serverId,
+                interpreter: kernel.kernelConnectionMetadata.interpreter
+            };
+            if (!this.registeredConnections.has(this.getControllerId(liveKernelConnection, 'jupyter-notebook'))) {
+                this.add(liveKernelConnection, [JupyterNotebookView, InteractiveWindowView]);
+            }
+            const liveController = Array.from(this.registeredControllers.values()).find(
+                (item) =>
+                    item.controller.notebookType === kernel.controller.notebookType &&
+                    item.connection.kind === 'connectToLiveRemoteKernel' &&
+                    item.connection.id === kernel.session?.kernel?.id
+            );
+            if (liveController) {
+                this.kernelProvider.updateKernel(kernel, liveController.connection, liveController.controller);
+
+                const notebookEditor = this.notebook.notebookEditors.find(
+                    (item) => item.notebook.uri.toString() === kernel.uri.toString()
+                );
+                await this.commandManager.executeCommand('notebook.selectKernel', {
```

Comment:
I'm confused about this - I thought the new proxy classes were being updated so that we wouldn't actually have to switch to a new kernel.

Reply:
We don't have that API yet, the plan is to work on that API with core soon, after we have this PR